File: internal/rules/no_ex_assign/no_ex_assign_test.go
package no_ex_assign

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/import/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoExAssignRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&NoExAssignRule,
		// Valid cases - ported from ESLint
		[]rule_tester.ValidTestCase{
			{Code: "try { } catch (e) { three = 2 + 1; }"},
			{Code: "try { } catch ({e}) { this.something = 2; }"},
			{Code: "function foo() { try { } catch (e) { return false; } }"},
		},
		// Invalid cases - ported from ESLint
		[]rule_tester.InvalidTestCase{
			{
				Code: "try { } catch (e) { e = 10; }",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 21},
				},
			},
			{
				Code: "try { } catch (ex) { ex = 10; }",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 22},
				},
			},
			{
				Code: "try { } catch (ex) { [ex] = []; }",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 22},
				},
			},
			{
				Code: "try { } catch (ex) { ({x: ex = 0} = {}); }",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 27},
				},
			},
			{
				Code: "try { } catch ({message}) { message = 10; }",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpected", Line: 1, Column: 29},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint/rules/no-ex-assign.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-ex-assign', {
  valid: [
    'try { } catch (e) { three = 2 + 1; }',
    'try { } catch ({e}) { this.something = 2; }',
    'function foo() { try { } catch (e) { return false; } }',
    'try { } catch (e) { let e = 10; e = 10; }',
    'try { } catch (e) { console.log({e}); }',
    'try { } catch (e) { foo({x: e}); }',
  ],
  invalid: [
    {
      code: 'try { } catch (e) { e = 10; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (ex) { ex = 10; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (ex) { [ex] = []; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (ex) { ({x: ex = 0} = {}); }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch ({message}) { message = 10; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (e) { for (;;) { e = 10; } }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (e) { for (e in obj) {} }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (e) { e += 1; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (e) { e++; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (e) { --e; }',
      errors: [{ messageId: 'unexpected' }],
    },
    {
      code: 'try { } catch (e) { for (e of arr) {} }',
      errors: [{ messageId: 'unexpected' }],
    },
  ],
});
